shared/src/main/scala/org/adridadou/openlaw/parser/template/variableTypes/ClauseType.scala

```diff
@@ -0,0 +1,119 @@
+    prepareClauseName(mappingParameter, executionResult).flatMap({
+      case Some(source) =>
+        Success(TemplateDefinition(name = source, path = path, mappingInternal = parameters.map({ case (key, value) => key.name -> value })))
+      case None =>
+        Failure("name cannot be resolved yet!")
+    })
+  }
+
+  override def access(value: Any, keys: Seq[String], executionResult: TemplateExecutionResult): Result[Any] = keys.toList match {
+    case Nil =>
+      Right(value)
+    case head::tail =>
+      executionResult.subExecutions.get(VariableName(head)).flatMap(subExecution =>
+        subExecution.getExpression(VariableName(head))
+          .flatMap(variable => variable.evaluate(subExecution)
+            .map(subValue => variable.expressionType(subExecution).access(subValue, tail, subExecution))
+          )) match {
+        case Some(result) =>
+          result
+        case None =>
+          throw new RuntimeException(s"properties '${tail.mkString(".")}' could not be resolved in sub clause '$head'")
+      }
+  }
+
+  override def keysType(keys: Seq[String], executionResult: TemplateExecutionResult): VariableType = {
+    keys.toList match {
+      case Nil =>
+        TemplateType
+      case head::tail =>
+        executionResult.subExecutions.get(VariableName(head)).flatMap(subExecution =>
+          subExecution.getExpression(VariableName(head))
+            .map(_.expressionType(subExecution).keysType(tail, subExecution))) match {
+          case Some(varType) =>
+            varType
+          case None =>
+            throw new RuntimeException(s"property '${tail.mkString(".")}' could not be resolved in sub clause '$head'")
+        }
+    }
+  }
+
+  override def validateKeys(name:VariableName, keys: Seq[String], executionResult: TemplateExecutionResult): Result[Unit] = keys.toList match {
+    case Nil =>
+      Success(())
+    case head::tail =>
+      executionResult.subExecutions.get(name).flatMap(subExecution =>
+        subExecution.getExpression(VariableName(head))
+          .map(variable => variable.expressionType(subExecution).keysType(tail,subExecution))) match {
+        case Some(_) =>
+          Success(())
+        case None =>
+          Failure(s"property '${keys.mkString(".")}' could not be resolved in sub clause '${name.name}'")
+      }
+  }
+
+  override def accessVariables(name:VariableName, keys:Seq[String], executionResult: TemplateExecutionResult): Seq[VariableName] = keys.toList match {
+    case Nil =>
+      Seq()
+    case head::tail =>
+      executionResult.subExecutions.get(name).flatMap(subExecution =>
+        subExecution.getExpression(VariableName(head))
+          .map(variable => variable
+            .expressionType(subExecution)
+            .accessVariables(VariableName(head), tail, subExecution)
+          )
+      ).getOrElse(Seq(name))
+  }*/
+
```

Comment:
The Clause Type should basically work like template but with a different behavior. What I propose here is to use Template Type to do almost everything (construct, cast etc ...) but we keep the Clause Type to tell the OpenlawExecutionEngine that the result needs to be embedded